### PR TITLE
Show all product combination images

### LINKED ISSUE #17317: [Improvement] Display all thumbnails for all combinations, only the main image is changed when selecting a combination
>Edit by Khouloud

**Improvement**: I need to show all variations images on the product page FO.
When the customer changes the combination, only the main image is changed.

There is a solution provided here: https://www.prestashop.com/forums/topic/963580-showing-thumbnails-for-all-combinations/

**Describe the bug**

BIG problem to correctly associate product color combinations with pictures without ruining customer experience and product page ergonomic...

Ex : lingerie product with different colors and size

https://www.sexy-cherry.com/en/p/43-634-amazing-sexy-panties-women-high-waist-lace-thongs-and-g-strings-underwear-ladies-hollow-out-underpants-intimates-lingerie.html#/8-color-white/30-size-xs

BO : 
> 1/ attribute type for size : radio, attribute type for color : color & texture
> 2/ create product combinations with two different attributes sizes and colors 
> 3/ associate each color combination to the right image (ex : all the white lingerie combinations (S,M,L,XL) will be associated with the white lingerie picture... and so on for each color combination)

FO Product page :

- there is always one little 'slide/carousel' near the main big image (either on the left or under the main product picture). This slide is supposed to show ALL the product variants while the big image shows the customer's choice.
- if i strictly associate each image to each color, the product page won't show any other variant picture than the one chosen as 'default combination' 
- if tweak a bit (associate 'incorrectly' default combination with ALL the pictures if i want to show all variant on the 'landing' product page), then if i click on the desired color swatch (ex : red) ALL the variants pictures (ex : white, black, blue...) will disappear in the slide/carousel !! 
- If the customer want to see other variants, he cannot use the 'mini slide/carousel, he is obliged to click on the color swatches to see the other colors of the same product

FO Catalog / Carousels :

- the images doesn't fade/change when i hover on them
- i have to tweak  (associate 'incorrectly' default combination with ALL the pictures) to have the fading effect in the product box

Needless to say : VERY BAD for sales

State of art and functionalities : any lingerie/mode/clothes shop, you'll see ALL products color variations ALWAYS visible near the main image on the product page...

Tested on
PrestaShop version: 1.7.5.1 - 1.7.6.2
PHP version: 7.2

### PR DESCRIPTION
**Pull request by MarkALeonard**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Show all images for a product and select the current combination image instead of only showing combination images. Also updates catalog product miniature to use new  catalogCover image so that all sizes and legend are available. Removes need to call Product::getCover() and $link->getImageLink() from the template.
| Type?         | improvement
| Category?     | FO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #17317
| How to test?  | Add additional images to a product.<br>Select specific images for a combination. <br>View the product in FO, all product images should be shown.<br> Change a combination and the combination image should be selected in the thumbnails and shown as the main image. <br>On the catalog listing the selected cover should be displayed and the correct alt atrribute assigned. 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17943)

### CODE REVIEW

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]
Front looks fine for me

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 1]
```diff
@@ -603,10 +603,29 @@ private function fillImages(
         if (!isset($this->product['cover'])) {
             if (count($this->product['images']) > 0) {
                 $this->product['cover'] = array_values($this->product['images'])[0];
+                foreach ($this->product['images'] as $image) {
+                    if (isset($image['cover']) && null !== $image['cover']) {
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]:
```suggestion
                    if (isset($image['cover'])) {
```
It's the same thing

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 2]
```diff
@@ -603,10 +603,29 @@ private function fillImages(
...
+                    }
+                }
             } else {
                 $this->product['cover'] = null;
             }
         }
+
+        if (count($this->product['images']) > 0) {
+            foreach ($this->product['images'] as $image) {
+                if (isset($image['cover']) && null !== $image['cover']) {
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]:
```suggestion
                if (isset($image['cover'])]) {
```

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Little thing for performance

**Review by Progi1984 — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @MarkALeonard,

This PR is OK but we need to confirm this behavior from our Product team.
Ping @PrestaShop/prestashop-product-team what do you think?

**Without PR**
We have this behavior
image

**With PR**, we have all images are displayed
image

**With this PR**, if we have more than one picture selected to a combination, we cannot distinguish which the second image for the combination selected
`For example, in the BO`
image

`In the FO`
image

Waiting for PM validation.

Thanks!

**Comment by colinegin** [comment]
Hello @MarkALeonard , thank you for your PR.
To be honest, I find it a bit weird for the user experience. It means that when looking at the different photos available for the product, it will automatically change the selected combination ? What if the combination is not available ? The photo is not displayed ? 
Wdyt @TristanLDD @prestascott ?

**Comment by prestascott** [comment]
@MarkALeonard Thank you for your PR,

In my opinion, images are shown depending on the combination. I agree with @colinegin, we can't show 2 images with different colors of t-shirts (black and white) when a specific combination is selected (white t-shirt). Selecting colors shouldn't be done in the image gallery but only on the color section. In order to know which image appears, we need to keep the indication when the image is selected like we currently have in the FO.

Here are two screenshots of a t-shirt product page from Mango website:
Capture d’écran 2020-04-20 à 12 00 57
Capture d’écran 2020-04-20 à 12 01 05

**Comment by matks** [comment]
It looks like this PR got stale following UX feedback. I consequently close it. Feel free to open it again if you come back 😉

**Comment by theodiablo** [comment]
This is exactly what I was looking for thanks a lot!

I am switching from Woocommerce, and this was the default dehaviour.

I think that when you have multiple photo per each variation, this feature might not be necessary, but if you only have one picture, you might want to display all images to fill a little the screen and show that there is more than one photo.

It would be awesome if this could be part of the core...

**Comment by vulpeep** [comment]
> Here are two screenshots of a t-shirt product page from Mango website

On the Mango website, this is justified by the fact that they provide several photos of the item from different angles for each colour. Moreover, clothing brands like Mango, H\&M, Zara, Bershka display the product photo in full size right away, without a thumbnails block, which doesn't match the standard Prestashop UI. In my shop, when there's only one photo per colour, a thumbnails block with just one icon looks ridiculous. That's why I'd like to be able to show all combinations in the thumbnails. It doesn't have to be the default logic - a toggle would do.

- **Aliexpress** displays images of all combinations straight away.
- **Ebay** displays images of all combinations straight away. Additionally, when a corresponding thumbnail is selected, the product variant switches automatically.
- **Uniqlo** displays several general images for all combinations plus one image for the specifically selected combination.

**Pull request closed without merging**